csharp/src/Google.Protobuf/Reflection/FileDescriptor.cs

```diff
@@ -445,9 +446,9 @@ private void CrossLink()
             }
         }
 
-        private static void AddAllExtensions(FileDescriptor[] dependencies, GeneratedClrTypeInfo generatedInfo, ExtensionRegistry registry)
+        private static IEnumerable<Extension> GetAllExtensions(FileDescriptor[] dependencies, GeneratedClrTypeInfo generatedInfo)
         {
-            registry.AddRange(dependencies.SelectMany(GetAllDependedExtensions).Concat(GetAllGeneratedExtensions(generatedInfo)).ToArray());
+            return dependencies.SelectMany(GetAllDependedExtensions).Distinct().Concat(GetAllGeneratedExtensions(generatedInfo));
```

Comment:
it feels that using Distinct() over Extension objects which do not explicitly define any equality / hashcode rules is fragile.

Reply:
In ExtensionRegistry, the identity of an extension is the pair (extension.TargetType, extension.FieldNumber), so we should honor that when de-duplicating. - looks like you could use Distinct overload that takes `IEqualityComparer`?

Reply:
So should we do Distinct over the file descriptors instead like above? Either way we're relying on reference equality which I think should be fine.

Reply:
I have found the same problem in ```GetAllGeneratedExtensions(generatedInfo)``` when two or more depended proto files import the same BASE proto with extensions. So i push another PR(#6954) to distinct the final collection and add a unit test with a more complicated situations.

Reply:
There's another issue (#6956) that I believe can be fixed by recursively gathering file descriptors from all dependencies. Can I go ahead and add the fix to this PR with tests @jtattermusch?

Reply:
Actually the issue may be unrelated depending on what James says. But it does describe another issue: extensions don't support dynamic loading.

Reply:
Thanks, you are right. I modify my patch and just collect and distinct all depended FileDescriptors for my usage. I'm not very familar with C#. If you have better ways to do this I will close my PR(#6954) .

Reply:
@ObsidianMinor  we still haven't addressed my original concern (using Distinct without a well defined equality comparer).